RF Correlate allows calculating pairwise correlations of structure probing experiments. It can be invoked either on individual transcripts or on whole XML folders, as well as RC files.<br/>
Overall, as well as per-transcript correlations are reported in TSV format.

Since v2.9.6, RF Correlate can calculate correlations between any number of experiments (many vs. many), and it can handle genome-level RC files (such as those generated by `rf-count-genome`).

!!! note "Note"
    When directly comparing two XML files, no check is made on the transcript ID, hence allowing the direct correlation of any two XML files (provided that they share the same sequence).
<br/>

# Usage

```bash
$ rf-correlate [options] sample1.rc sample2.rc .. sampleN.rc
$ rf-correlate [options] rna1.xml rna2.xml .. rnaN.xml
$ rf-correlate [options] XML_dir_1/ XML_dir_2/ .. XML_dir_N/
```

To list all available parameters, simply type:

```bash
$ rf-correlate -h
```

Parameter         | Type | Description
----------------: | :--: |:------------
__-p__ *or* __--processors__ | int | Number of processors to use (Default: __1__)
__-o__ *or* __--output__ | string | Output folder (Default: __rf_correlate/__)
__-ow__ *or* __--overwrite__ | | Overwrites output folder (if the specified folder already exists)
__-m__ *or* __--min-values__ | float | Minimum number of values to calculate correlation (Default: __off__)<br/>__Note:__ if a value between 0 and 1 is provided, this is interpreted as a fraction of the transcript's length 
__-cr__ *or* __--cap-react__ | float | Maximum reactivity value to cap reactivities to (&gt; 0, Default: __1e9__)<br/>__Note:__ if processing RC files, this parameter only applies to ratios (`-r`)
__-mr__ *or* __--max-react__ | float |  Reactivity values above this threshold will be excluded from correlation calculation (&gt; 0, Default: __none__)<br/>__Note:__ if processing RC files, this parameter only applies to ratios (`-r`)
__-I__ *or* __--ingore-sequence__ | | Ignores sequence differences (e.g. SNVs) between the compared transcripts
__-S__ *or* __--spearman__ | | Uses Spearman instead of Pearson to calculate correlation
 | | __RC file-specific options__
__-i__ *or* __--index__ | string | An RCI index file to be used for all input RC files<br/>__Note:__ If no RCI index is provided, RF Correlate will look for files with .rci extension in the same input folder as the RC files, named after the RC files (e.g., Sample.rc will look for Sample.rc.rci). If no RCI file is found, it will be created at runtime, and stored in the same folder of the input RC files.
__--kb__ *or* __--keep-bases__ | string | Bases on which correlation should be calculated (Default: __all__)<br/>__Note:__ this option has effect only on RC files. For XML files, reactive bases are automatically identified from the ``reactive`` attribute
__-mc__ *or* __--min-coverage__ | int | Restricts the correlation analysis to bases exceeding this coverage
__-c__ *or* __--coverage__ | | Correlation is calculated on the coverage, rather than on the raw RT stop/mutation counts
__-r__ *or* __--ratio__ | | Correlation is calculated on the ratio between, the RT stop/mutation counts and the coverage, rather than on the raw RT stop/mutation counts
__-bs__ *or* __--block-size__ | int | Defines the size of the memory block (in bp) to process RC files containing whole chromosome data (such as those generated by rf-count-genome) (&ge;1, Default: __100000__)


!!! note "Note"
    When ``--min-values`` specified value is interpreted as a fraction of the transcript's length, only reactive bases (specified by the ``reactive`` [attribute](https://rnaframework-docs.readthedocs.io/en/latest/rf-norm/) for XML files, or via the ``--keep-bases`` parameter for RC files) are considered. For example, if a transcript containing 25% of each base has been modified with DMS (than only modifies A/C residues), setting ``--min-values`` to 0.5 will cause RF Correlate to skip the transcript if more than 50% of the A/C residues are NaNs (or do not exceed the ``--min-coverage`` threshold for RC files).
    
<br/>
## Sample labeling
By default, RF Correlate uses the input file names (stripped of their extension) as labels in the output files.

It is however possible to specify custom labels by prepending them to the input files, in the form `label:`:

```bash
$ rf-correlate [options] Sample_1:XML_dir_1/ Sample_2:XML_dir_2/ .. Sample_N:XML_dir_N/
```   

<br/>
## Output files
RF Correlate generates an output folder containing:

1. __matrix.csv__: a CSV matrix of the pairwise correlation coefficients between all samples
2. __pairwise/__: a folder containing a TSV file for each pairwise comparison, with a list of correlation coefficients (and corresponding p-values) for each transcript (or genomic block)
3. __heatmap.pdf__ (*optional*, requires `-g`): a clustered heatmap of the pairwise correlation coefficients between all samples

<br/>
![RF Correlate Heatmap](http://www.incarnatolab.com/images/docs/RNAframework/rf-correlate_heatmap.png)