The RF Norm module takes one (Rouskin and Zubradt methods), two (Ding and Siegfried methods), or three (Siegfried method) RC files generated by the RF Count module, and performs normalization to obtain transcriptome-wide per-base reactivities.<br/>Reactivity scores can be computed using 3 methods:
<br/><br/>
## Scoring of RT-stops/nuclease cuts-based methods
__[1] Ding *et al*., 2014 (PMID: [24270811](https://www.ncbi.nlm.nih.gov/pubmed/24270811))__<br/><br/>
Per-base signal is calculated as the ratio of the natural log (ln) of the raw count of RT-stops/nuclease cuts at a given position of a transcript, to the average of the ln of RT-stops/nuclease cuts along the whole transcript:<br/>

<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><msub><mi>U</mi><mi>i</mi></msub><mo>=</mo><mfrac><mrow><mi>ln</mi><mo>(</mo><msub><mi>n</mi><mrow><mn>U</mn><mi>i</mi></mrow></msub><mo>+</mo><mi>p</mi><mo>)</mo></mrow><mfenced><mrow><munderover><mo>&#x2211;</mo><mrow><mi>j</mi><mo>=</mo><mn>0</mn></mrow><mi>l</mi></munderover><mstyle displaystyle="true"><mfrac><mrow><mi>ln</mi><mo>(</mo><msub><mi>n</mi><mrow><mn>U</mn><mi>j</mi></mrow></msub><mo>+</mo><mi>p</mi><mo>)</mo></mrow><mi>l</mi></mfrac></mstyle></mrow></mfenced></mfrac></math>
<br/>
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><msub><mi>T</mi><mi>i</mi></msub><mo>=</mo><mfrac><mrow><mi>ln</mi><mo>(</mo><msub><mi>n</mi><mrow><mn>T</mn><mi>i</mi></mrow></msub><mo>+</mo><mi>p</mi><mo>)</mo></mrow><mfenced><mrow><munderover><mo>&#x2211;</mo><mrow><mi>j</mi><mo>=</mo><mn>0</mn></mrow><mi>l</mi></munderover><mstyle displaystyle="true"><mfrac><mrow><mi>ln</mi><mo>(</mo><msub><mi>n</mi><mrow><mn>T</mn><mi>j</mi></mrow></msub><mo>+</mo><mi>p</mi><mo>)</mo></mrow><mi>l</mi></mfrac></mstyle></mrow></mfenced></mfrac></math>
<br/>
where *n<sub>Ui</sub>* and *n<sub>Ti</sub>* are respectively the raw read counts in the untreated (or RNase V1) and treated (DMS, CMCT, SHAPE, or Nuclease S1) samples at position *i* of the transcript, *l* is the transcript’s length, and *p* is a pseudocount added to deal with non-covered regions. *U<sub>i</sub>* and *T<sub>i</sub>* are respectively the normalized number of RT-stops/nuclease cuts at position *i* in the untreated and treated samples.<br/>Score at position i is then calculated as:<br/><br/>
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><msub><mi>S</mi><mi>i</mi></msub><mo>=</mo><mi>m</mi><mi>a</mi><mi>x</mi><mo>(</mo><mn>0</mn><mo>,</mo><mo>&#xA0;</mo><mo>(</mo><msub><mi>T</mi><mi>i</mi></msub><mo>-</mo><msub><mi>U</mi><mi>i</mi></msub><mo>)</mo><mo>)</mo></math>
<br/>
__[2] Rouskin *et al*., 2014 (PMID: [24336214](https://www.ncbi.nlm.nih.gov/pubmed/24336214))__<br/><br/>
The untreated sample is not considered. Per-base RT-stops/nuclease cuts are used as a direct measure of the raw signal.

!!! warning "Warning"
    Normalization of data processed by Rouskin method, can only be performed using the __90% Winsorizing__ approach.
<br/>
## Scoring of mutational profiling-based methods
__[3] Siegfried *et al*., 2014 (PMID: [25028896](https://www.ncbi.nlm.nih.gov/pubmed/25028896))__<br/><br/>
This method takes into account both an untreated sample, and (optionally) a denatured control
sample.<br/>Per-base raw signal is calculated as:<br/><br/>

<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><msub><mi>S</mi><mi>i</mi></msub><mo>=</mo><mfrac><mrow><mstyle displaystyle="true"><mfrac><msub><mi>n</mi><mrow><mi>T</mi><mi>i</mi></mrow></msub><msub><mi>c</mi><mrow><mi>T</mi><mi>i</mi></mrow></msub></mfrac></mstyle><mo>-</mo><mstyle displaystyle="true"><mfrac><msub><mi>n</mi><mrow><mi>U</mi><mi>i</mi></mrow></msub><msub><mi>c</mi><mrow><mi>U</mi><mi>i</mi></mrow></msub></mfrac></mstyle></mrow><mstyle displaystyle="true"><mfrac><msub><mi>n</mi><mrow><mi>D</mi><mi>i</mi></mrow></msub><msub><mi>c</mi><mrow><mi>D</mi><mi>i</mi></mrow></msub></mfrac></mstyle></mfrac></math>
<br/>
where *n<sub>Ti</sub>*, *n<sub>Ui</sub>*, and *n<sub>Di</sub>* are respectively the mutation counts in the treated, untreated, and denatured samples at position *i* of the transcript, while *c<sub>Ti</sub>*, *c<sub>Ui</sub>*, and *c<sub>Di</sub>* are respectively the reads covering position *i* of the transcript in the treated, untreated, and denatured samples.<br/>If no denatured control sample is provided, raw reactivities are simply calculated as:<br/><br/>
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><msub><mi>S</mi><mi>i</mi></msub><mo>=</mo><mfrac><msub><mi>n</mi><mrow><mi>T</mi><mi>i</mi></mrow></msub><msub><mi>c</mi><mrow><mi>T</mi><mi>i</mi></mrow></msub></mfrac><mo>-</mo><mfrac><msub><mi>n</mi><mrow><mi>U</mi><mi>i</mi></mrow></msub><msub><mi>c</mi><mrow><mi>U</mi><mi>i</mi></mrow></msub></mfrac></math>
<br/><br/>
__[4] Zubradt *et al*., 2016 (PMID: [27819661](https://www.ncbi.nlm.nih.gov/pubmed/27819661))__<br/><br/>
The untreated sample is not considered. Per-base raw signal is calculated as:<br/><br/>
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><msub><mi>S</mi><mi>i</mi></msub><mo>=</mo><mfrac><msub><mi>n</mi><mrow><mi>T</mi><mi>i</mi></mrow></msub><msub><mi>c</mi><mrow><mi>T</mi><mi>i</mi></mrow></msub></mfrac></math>
<br/>
where *n<sub>Ti</sub>*, and *c<sub>Ti</sub>* are respectively the mutations count and the read coverage at position *i* of the transcript.
<br/><br/>
## Normalization of raw reactivities
Raw reactivity scores can be normalized using 3 different approaches:<br/><br/>

Method         | Description
-------------: | :------------
__2-8% Normalization__ | From the top 10% of values, the top 2% is ignored, then any reactivity value along the entire transcript is divided by the average of the remaining 8%
__90% Winsorizing__ | Each reactivity value above the 95<sup>th</sup> percentile is set to the 95<sup>th</sup> percentile and each reactivity value below the 5<sup>th</sup> percentile is set to the 5<sup>th</sup> percentile, then the reactivity at each position of the transcript is divided by the value of the 95<sup>th</sup> percentile
__Box-plot Normalization__ | Values greater than 1.5x the interquartile range (numerical distance between the 25<sup>th</sup> and 75<sup>th</sup> percentiles) above the 75<sup>th</sup> percentile are removed. After excluding these outliers, the next 10% of remaining reactivities are averaged, and all reactivities (including outliers) are divided by this value.

<br/>
Normalized reactivities can be further remapped to values ranging from 0 to 1 according to Zarringhalam *et al.*, 2012 (PMID: [23091593](https://www.ncbi.nlm.nih.gov/pubmed/23091593)). In this approach, values < 0.25 are linearly mapped to [0-0.35[, values &ge; 0.25 and < 0.3 are linearly mapped to [0.35-0.55[, values &ge; 0.3 and < 0.7 are linearly mapped to [0.55-0.85[, and values &ge; 0.7 are linearly mapped to [0.85-1].<br /><br />

## Sliding-window normalization
RF Norm supports data normalization in sliding windows. Windows can be both static (default) or dynamic. When a window size is chosen, data normalization is performed by sliding by the chosen offset, a window of that size. While choice of window's type is irrelevant with SHAPE data, it becomes particularly important when dealing with base-specific reagents. Let consider the example below, in which an RNA has been modified by DMS, which only reacts with A and C residues.<br/><br/>![Static vs. Dynamic windows](http://www.rnaframework.com/images/static_vs_dynamic_windows.png)<br/><br/>
In the above example, use of static windows of size __10 nt__ results in an erroneous overestimation of base reactivities for certain residues (marked in red). This is caused by the fact that A/C residues are unevenly distributed along the transcript, thus causing certain windows to have far less than 50% of A/C bases (contrary to what it would be expected by chance). Instead, use of dynamic windows of size __10 nt__ avoids this overestimation, as the window's size is dynamically adjusted to always include 10 A/C residues.</br> The overestimation effect can also be minimized by increasing the size of static windows.
<br/><br/>
# Usage
To list the required parameters, simply type:

```bash
$ rf-norm -h
```

Parameter         | Type | Description
----------------: | :--: |:------------
__-u__ *or* __--untreated__ | string | Path to the RC file for the non-treated/denatured (or RNase V1) sample <br/>(required by Ding/Siegfried scoring methods)
__-t__ *or* __--treated__ | string | Path to the RC file for the treated (or Nuclease S1) sample
__-d__ *or* __--denatured__ | string | Path to the RC file for the denatured sample <br/>(optional for Siegfried scoring method)
__-i__ *or* __--index__ | string[,string] | A comma separated (no spaces) list of RCI index files for the provided RC files<br/>__Note #1:__ RCI files must be provided in the order 1. Untreated/Denatured, 2. Treated<br/>__Note #2:__ If a single RTI file is specified, it will be used for all RC files<br/>__Note #3:__ If no RCI index is provided, it will be generated at runtime, and stored in the same folder of the untreated/denatured/treated samples
__-p__ *or* __--processors__ | int | Number of processors (threads) to use (Default: __1__)
__-o__ *or* __--output-dir__ | string | Output directory for writing normalized data in XML format (Default: __&lt;treated&gt;\_vs\_&lt;untreated&gt;\_norm/__ for Ding method or Siegfried method without denatured sample, __&lt;treated&gt;\_norm/__ for Rouskin/Zubradt methods, __&lt;treated&gt;\_vs\_&lt;untreated&gt;\_&lt;denatured&gt;\_norm/__ for Siegfried method with denatured sample)
__-ow__ *or* __--overwrite__ | | Overwrites the output directory if already exists
__-c__ *or* __--config-file__ | string | Path to a configuration file with normalization parameters (see *Configuration files* paragraph)<br/>__Note #1:__ If the provided file exists, the loaded configuration will override any command-line specified parameter<br/>__Note #2:__ If the provided file doesn’t exist, it will be generated using the specified command-line (or default) parameters
__-sm__ *or* __--scoring-method__ | int | Method for score calculation (1-4, Default: __1__):<br/>__1.__ Ding *et al.*, 2014 <br/>__2.__ Rouskin *et al.*, 2014 <br/>__3.__ Siegfried *et al.*, 2014<br/>__4.__ Zubradt *et al.*, 2016
__-nm__ *or* __--norm-method__ | int | Method for signal normalization (1-3, Default: __1__):<br/>__1.__ 2-8% Normalization <br/>__2.__ 90% Winsorizing <br/>__3.__ Box-plot Normalization
__-r__ *or* __--raw__ | | Reports raw reactivities (skips data normalization)
__-rm__ *or* __--remap-reactivities__ | | Remaps normalized reactivities to values ranging from 0 to 1 according to Zarringhalam *et al*., 2012
__-rb__ *or* __--reactive-bases__ | string | Reactive bases to consider for signal normalization (Default: __N__ [ACGT])<br/>__Note:__ This parameter accepts any IUPAC code, or their combination (e.g. ``-rb M``, or ``-rb AC``). Any other base will be reported as NaN
__-ni__ *or* __--norm-independent__ | | Each one of the reactive bases will be normalized independently (e.g. -rb AC -ni will independently normalize A and C residues)
__-dw__ *or* __--dynamic-window__ | | When enabled, the normalization window is dynamically resized to include at least that number of reactive bases (e.g. ``-rb AC -nw 50 -dw`` instructs RF Norm to normalize reactivities in windows containing at least 50 A/C residues)
__-mc__ *or* __--mean-coverage__ | float | Discards any transcript with mean coverage below this threshold (&ge;0, Default: __0__)
__-ec__ *or* __--median-coverage__ | float | Discards any transcript with median coverage below this threshold (&ge;0, Default: __0__)
__-nw__ *or* __--norm-window__ | int | Window size (in nt) for signal normalization (&ge;3, Default: __whole transcript__ [Ding; Siegfried], __50__ [Rouskin; Zubradt])<br/>__Note:__ a maximum window size of 30,000 nt is allowed when ``-dw`` (or ``--dynamic-window``) is enabled
__-wo__ *or* __--window-offset__ | int | Offset (in nt) for window sliding during normalization (Default: __none__ [Ding; Siegfried], __50__ [Rouskin; Zubradt])
__-D__ *or* __--decimals__ | int | Number of decimals for reporting reactivities (1-10, Default: __3__)
__-n__ *or* __--nan__ | int | Positions of transcript with read coverage behind this threshold, will be reported as NaN in the reactivity profile (&gt;0, Default: __10__)
 | | __Scoring method #1 options (Ding *et al*., 2014)__
__-pc__ *or* __--pseudocount__ | float | Pseudocount added to reactivities to avoid division by 0 (&gt;0, Default: __1__)
__-s__ *or* __--max-score__ | float | Score threshold for capping raw reactivities (&gt;0, Default: __10__)
 | | __Scoring method #3 options (Siegfried *et al*., 2014)__
__-mu__ *or* __--max-untreated-mut__ | float | Maximum per-base mutation rate in untreated sample (&le;1, Default: __0.05__ [5%])
__-mm__ *or* __--max-mutation-rate__ | float | Maximum per-base mutation rate (&le;1, Default: __0.2__ [20%])
 | | __Scoring method #4 options (Zubradt *et al*., 2017)__
__-mm__ *or* __--max-mutation-rate__ | float | Maximum per-base mutation rate (&le;1, Default: __0.2__ [20%])

<br/>
## Configuration files
RF Norm configuration files are used to provide normalization parameters for the analysis, without the need to manually specify them from the command-line.<br/>Configuration files are composed of a list of key/value pairs, separated by the equal sign (=), or by the colon punctuation mark (:). Keys and values are case-insensitive.<br/>
Accepted key/value pairs are:<br/>

Parameters     | Accepted values | Default value
-------------: | :------------:  | :------------
__scoreMethod__ | "Ding" (or 1); "Rouskin" (or 2); "Siegfried" (or 3); "Zubradt" (or 4) | __Ding__
__normMethod__ | "2-8%" (or 1); "90% Winsorizing" (or 2); "Box-plot" (or 3) | __2-8%__
__reactiveBases__ | \[ACGTURYSWKMBDHVN\] (or "all") | __all__
__normIndependent__ | TRUE/FALSE; Yes/No; 1/0 | __FALSE__
__normWindow__ | Positive integer &ge; 3 | __1e9__ [Ding; Siegfried]<br/>__50__ [Rouskin; Zubradt]
__windowOffset__ | Positive integer &gt; 0 | __1e9__ [Ding; Siegfried]<br/>__50__ [Rouskin; Zubradt]
__meanCoverage__ | Positive float &ge; 0 | __0__
__medianCoverage__ | Positive float &ge; 0 | __0__
__remapReactivities__ | TRUE/FALSE; Yes/No; 1/0 | __FALSE__
 | __Scoring method #1 options__ | 
__maxScore__ | Positive float &gt; 0 | __10__
__pseudoCount__ | Positive float &gt; 0 | __1__
 | __Scoring method #3 options__ | 
__maxUntreatedMut__ | 0 &le; r &le; 1 | __0.05__
__maxMutationRate__ | 0 &le; r &le; 1 | __0.2__


```bash
# A sample configuration file

scoreMethod=Ding
normMethod=2-8%
maxScore=10
pseudoCount=1
reactiveBases=N
normIndependent=FALSE
normWindow=1e9
windowOffset=1e9
meanCoverage=1
```

<br/>
## Output XML files
RF Norm produces a XML file for each transcript being analyzed, with the following structure:<br/>

```xml
<?xml version="1.0" encoding="UTF-8"?><data [attributes]>	<transcript id=”Transcript ID” length=”Transcript length”>		<sequence>			Transcript sequence		</sequence>		<reactivity>			Comma-separated list of reactivity values		</reactivity>	</transcript></data>
```
The __data__ tag’s attributes allow keeping track of the analysis performed:<br/>

Attribute     | Possible values | Description
-------------: | :------------: | :----------
__tool__ | rf-norm | The tool that generated this XML file
__scoring__ | Ding, Rouskin, Siegfried, or Zubradt | Scoring method
__norm__ | 2-8%, Winsorizing 90%, or Box-plot | Normalization method
__reactive__ | \[ACGT\] | Reactive bases
__win__ | Positive integer &ge; 3 | Normalization window's size (in nt)
__offset__ | Positive integer &ge; 1 | Offset for normalization window sliding
__remap__ | TRUE/FALSE | Whether normalized reactivities have been remapped according to Zarringhalam *et al*., 2012
 | | __Scoring method #1 (Ding *et al*., 2014)__
__max__ | Positive float &gt; 0 | Score threshold for capping raw reactivities
__pseudo__ | Positive float &gt; 0 | Pseudocount added to avoid division by 0 during reactivity calculation
 | | __Scoring method #3 (Siegfried *et al*., 2014)__
__maxumut__ | 0 &le; r &le; 1 | Maximum per-base mutation rate in untreated sample
__maxmutrate__ | 0 &le; r &le; 1 | Maximum per-base mutation rate
 | | __Scoring method #4 (Zubradt *et al*., 2017)__
__maxmutrate__ | 0 &le; r &le; 1 | Maximum per-base mutation rate