The RF PeakCall module takes two RC files generated by the RF Count module, and performs transcriptome-wide peak calling from RNA immunoprecipitation (IP) experiments.<br/>
Analysis is performed by sliding a window of length *w* along the transcript, and calculating the signal enrichment in the IP sample versus control sample as:<br/>

<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><msub><mi>E</mi><mrow><mo>(</mo><mi>i</mi><mo>.</mo><mo>.</mo><mi>i</mi><mo>+</mo><mi>w</mi><mo>)</mo></mrow></msub><mo>=</mo><mfrac bevelled="true"><mrow><munderover><mo>&#x2211;</mo><mrow><mi>j</mi><mo>=</mo><mn>i</mn></mrow><mrow><mi>i</mi><mo>+</mo><mi>w</mi></mrow></munderover><msub><mi>log</mi><mrow><mo>2</mo></mrow></msub><mfenced><mfrac><mstyle displaystyle="true"><mfrac bevelled="true"><mfenced><mrow><msub><mi>&#x3BC;</mi><mrow><mi>I</mi><mi>P</mi><mo>(</mo><mi>j</mi><mo>)</mo></mrow></msub><mo>+</mo><mi>p</mi></mrow></mfenced><mfenced><mrow><mi>M</mi><msub><mi>d</mi><mrow><mi>I</mi><mi>P</mi></mrow></msub><mo>+</mo><mi>p</mi></mrow></mfenced></mfrac></mstyle><mstyle displaystyle="true"><mfrac bevelled="true"><mfenced><mrow><msub><mi>&#x3BC;</mi><mrow><mi>C</mi><mi>t</mi><mi>r</mi><mi>l</mi><mo>(</mo><mi>j</mi><mo>)</mo></mrow></msub><mo>+</mo><mi>p</mi></mrow></mfenced><mfenced><mrow><mi>M</mi><msub><mi>d</mi><mrow><mi>C</mi><mi>t</mi><mi>r</mi><mi>l</mi></mrow></msub><mo>+</mo><mi>p</mi></mrow></mfenced></mfrac></mstyle></mfrac></mfenced></mrow><mrow><mi>w</mi></mrow></mfrac></math>
<br/>
where *i* and *i+w* are the start and end position of the window, *&#x3BC;<sub>IP(i)</sub>* and *&#x3BC;<sub>Ctrl(i)</sub>* are respectively the  coverage at position *i* in the IP and control samples, *Md<sub>IP</sub>* and *Md<sub>Ctrl</sub>* are respectively the median coverage on the whole transcript in the IP and control samples, and *p* is a pseudocount added to deal with non-covered regions/transcripts.<br/>When a control sample is not provided, the signal enrichment is simply calculated as:<br/>

<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><msub><mi>E</mi><mrow><mo>(</mo><mi>i</mi><mo>.</mo><mo>.</mo><mi>i</mi><mo>+</mo><mi>w</mi><mo>)</mo></mrow></msub><mo>=</mo><mfrac bevelled="true"><mrow><munderover><mo>&#x2211;</mo><mrow><mi>j</mi><mo>=</mo><mn>i</mn></mrow><mrow><mi>i</mi><mo>+</mo><mi>w</mi></mrow></munderover><msub><mi>log</mi><mrow><mo>2</mo></mrow></msub><mfenced><mstyle displaystyle="true"><mfrac bevelled="true"><mfenced><mrow><msub><mi>&#x3BC;</mi><mrow><mi>I</mi><mi>P</mi><mo>(</mo><mi>j</mi><mo>)</mo></mrow></msub><mo>+</mo><mi>p</mi></mrow></mfenced><mfenced><mrow><mi>M</mi><msub><mi>d</mi><mrow><mi>I</mi><mi>P</mi></mrow></msub><mo>+</mo><mi>p</mi></mrow></mfenced></mfrac></mstyle></mfenced></mrow><mrow><mi>w</mi></mrow></mfrac></math><br/>
A p-value is then calculated for each window with detected enrichment above a defined cutoff, using a __Fisher's exact test__. Thus, the following 2x2 contingency matrix is defined for each cutoff-passing window:<br/>

 &nbsp; | n<sub>11</sub> | n<sub>12</sub>
-------------: | :------------:  | :------------:
__n<sub>21</sub>__ | *&#x3BC;<sub>IP(i..i+w)</sub>* | *Md<sub>IP</sub>*
__n<sub>22</sub>__ | *&#x3BC;<sub>Ctrl(i..i+w)</sub>* | *Md<sub>Ctrl</sub>*

If no control sample is provided, the contingency matrix is instead defined as:<br/>

 &nbsp; | n<sub>11</sub> | n<sub>12</sub>
-------------: | :------------:  | :------------:
__n<sub>21</sub>__ | *&#x3BC;<sub>IP(i..i+w)</sub>* | *Md<sub>IP</sub>*
__n<sub>22</sub>__ | *&#x3BC;<sub>IP windows</sub>* | *Md<sub>IP</sub>*

where *&#x3BC;<sub>IP windows</sub>* is the average of the mean values for each possible window in the IP sample.<br/>
P-values are then subjected to Benjamini-Hochberg correction. Consecutive significantly enriched windows are then merged together, and p-values are combined by Stouffer's method.<br/>

## Peak refinement
Peaks are called by sliding a window of a fixed size, with a user-defined offset. This means that, in some cases, the called interval might not include the full peak (or it might include additional low-enrichment bases). Consider the following example:<br/><br/>
![Peak refinement](http://www.rnaframework.com/images/peak_refinement.png)
<br/><br/>
In this example, peaks are called using a window size of 150 nt (default offset: 75 nt) and an enrichment threshold of 3 fold. With default parameters, the called interval does not entangle the full peak (as the coverage slightly drops around half of the peak in the IP sample). When the ``-r`` (or ``--refine``) parameter is specified, RF PeakCall will progressively enlarge (or shrink) the interval to only include bases exceeding the enrichment threshold. In this case, refinement trims all the non-enriched bases from the 5' end of the interval. However, because of the slightly lower coverage of the second half of the peak, the enrichment (2.87) is just below the acceptance threshold. To account for such situations, the ``-x`` (or ``--relaxed``) parameter can be specified to enable a more *loose* evaluation of the coverage in the IP sample. Specifically, the coverage will be rounded to nearest multiple of 0.5.

!!! note "Note"
    Peak refinement also takes into account the median transcript coverage of the IP sample. While extending peak boundaries, only bases exceeding the median transcript coverage in the IP sample will be included, independently of their enrichment.
<br/>

# Usage
To list the required parameters, simply type:

```bash
$ rf-peakcall -h
```

Parameter         | Type | Description
----------------: | :--: |:------------
__-c__ *or* __--control__ | string | Path to the RC file for the control sample
__-I__ *or* __--IP__ | string | Path to the RC file for the immunoprecipitated (IP) sample
__-i__ *or* __--index__ | string[,string] | A comma separated (no spaces) list of RCI index files for the provided RC files<br/>__Note #1:__ RCI files must be provided in the order 1. Control, 2. IP<br/>__Note #2:__ If a single RTI file is specified, it will be used for all RC files<br/>__Note #3:__ If no RCI index is provided, it will be generated at runtime, and stored in the same folder of the control/IP samples
__-p__ *or* __--processors__ | int | Number of processors (threads) to use (Default: __1__)
__-o__ *or* __--output__ | string | Output folder (Default: __&lt;IP&gt;\_vs\_&lt;Control&gt;/__ if a control RC file is provided, or __&lt;IP&gt;/__ if only the IP RC file is provided)
__-ow__ *or* __--overwrite__ | | Overwrites the output folder (if the specified folder already exists)
__-g__ *or* __--img__ | | Enables the generation of coverage plots
__-mp__ *or* __--meta-plot__ | | Enables the generation of meta-gene plots (both coverage and peaks' distribution)
__-eo__ *or* __--plot-enriched-only__ | | Meta-gene plots will be calculated only on transcripts with a detected enrichment (requires ``-mp``)
__-w__ *or* __--window__ | int | Window size (in nt) for peak calling (&ge;10, Default: __150__)
__-f__ *or* __--offset__ | int | Offset (in nt) for window sliding (&ge;1, Default: __window / 2__)
__-md__ *or* __--merge-distance__ | int | Maximum distance (in nt) for merging non-overlapping windows (&ge;0, Default: __50__)
__-e__ *or* __--enrichment__ | float | Minimum log<sub>2</sub> enrichment in IP vs. Control for reporting a peak (&ge;1, Default: __3__)
__-v__ *or* __--p-value__ | float | P-value cutoff for reporting a peak (0 &le; *p* &le; 1, Default: __0.05__)
__-s__ *or* __--summit__ | | Generates an additional BED file containing the coordinates of peak summits (peak regions with the highest coverage)
__-r__ *or* __--refine__ | | Refines peak boundaries
__-x__ *or* __--relaxed__ | | Uses more relaxed criteria to refine peak boundaries (requires ``-r``)
__-pc__ *or* __--pseudocount__ | float | Pseudocount added to read counts to avoid division by 0 (&gt;0, Default: __1__)
__-mc__ *or* __--mean-coverage__ | float | Discards any transcript with mean coverage in control sample below this threshold (&ge;0, Default: __0__)
__-ec__ *or* __--median-coverage__ | float | Discards any transcript with median coverage in control sample below this threshold (&ge;0, Default: __0__)
__-D__ *or* __--decimals__ | int | Number of decimals for reporting enrichment/p-value (1-10, Default: __3__)

<br/>
## Output BED files
The peaks BED file contains 5 tab-delimited fields:<br/>


Field    | Description
-------------: | :----------
__Transcript ID__ | ID of the transcript
__Start__ | Start coordinate of the peak (0-based)
__End__ | End coordinate of the peak (0-based)
__Enrichment__ | Fold enrichment of the IP signal versus Control signal
__p-value__ | Benjamini-Hochberg adjusted p-value

<br/>
The summits BED file only contains 3 tab-delimited fields:<br/>


Field    | Description
-------------: | :----------
__Transcript ID__ | ID of the transcript
__Start__ | Start coordinate of the summit (0-based)
__End__ | End coordinate of the summit (0-based)
__Enrichment__ | Fold enrichment of the IP signal versus Control signal
__p-value__ | Benjamini-Hochberg adjusted p-value